src/modules/roc_audio/profiler.cpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_audio/profiler.h"
+#include "roc_core/log.h"
+#include "roc_core/panic.h"
+#include "roc_core/shared_ptr.h"
+
+namespace roc {
+namespace audio {
+
+Profiler::Profiler(core::IAllocator& allocator,
+                   packet::channel_mask_t channels,
+                   size_t sample_rate,
+                   core::nanoseconds_t interval)
+    : rate_limiter_(interval)
+    , interval_(interval)
+    , chunk_length_(sample_rate / 100)
+    , num_chunks_((size_t)(interval / (core::Second / 100)) + 1)
+    , chunks_(allocator)
+    , moving_avg_(0)
+    , sample_rate_(sample_rate)
+    , num_channels_(packet::num_channels(channels)) {
+    if (num_channels_ == 0) {
+        roc_panic("profiler: n_channels is zero");
+    }
+    if (sample_rate_ == 0) {
+        roc_panic("profiler: sample_rate is zero");
+    }
+    chunks_.resize(num_chunks_);
+}
+
+void Profiler::begin_frame(size_t frame_size) {
+    if (frame_size % num_channels_ != 0) {
+        roc_panic("profiler: unexpected frame size");
+    }
+}
```

Comment:
It seems we don't need this method anymore. Maybe move the check to end_frame() and remove begin_frame()? We can then rename end_frame() to add_frame() or something like this.